### PR TITLE
fix(http): ignore question mark when params are parsed

### LINKED ISSUE #28722: HttpParams fromString first param not updating
# 🐞 bug report

### Affected Package
@angular/common/http

### Description

The delete and set methods do not work as expected for the first param when fromString starts with a `?` i.e.

```
new HttpParams({ fromString: '?test=string' }).set('test', 'value').toString() // ?test=string&test=value instead of ?test=value
```
It's useful to support `?` because `window.location.search` starts with that character.

## 🔬 Minimal Reproduction
https://stackblitz.com/edit/angular-zztqky?file=src%2Fapp%2Fapp.component.ts

## 🌍  Your Environment

**Angular Version:**
Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.13.1
@angular-devkit/build-angular     0.13.1
@angular-devkit/build-optimizer   0.13.1
@angular-devkit/build-webpack     0.13.1
@angular-devkit/core              7.3.1
@angular-devkit/schematics        7.3.1
@angular/cli                      7.3.1
@ngtools/webpack                  7.3.1
@schematics/angular               7.3.1
@schematics/update                0.13.1
rxjs                              6.4.0
typescript                        3.2.4
webpack                           4.29.0

### PR DESCRIPTION
**Pull request by Jefiozie**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [X] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Fixes #28722

When the HTTP params are parsed it will not ignore the question mark params when this is in the string.

## What is the new behavior?

With this change we will replace/remove the `?` so that we won't have the problem as described in the issue.

## Does this PR introduce a breaking change?

- [ ] Yes
- [ X] No

### CODE REVIEW

File: packages/common/http/src/params.ts [hunk 1]
```diff
@@ -73,7 +73,7 @@ export class HttpUrlEncodingCodec implements HttpParameterCodec {
 function paramParser(rawParams: string, codec: HttpParameterCodec): Map<string, string[]> {
   const map = new Map<string, string[]>();
   if (rawParams.length > 0) {
-    const params: string[] = rawParams.split('&');
+    const params: string[] = rawParams.replace('?', '').split('&');
```
Comment (petebacondarwin) [inline packages/common/http/src/params.ts:76]: I think to be safe, we should only strip the `?` from the start of the string yes?
So

```suggestion
    const params: string[] = rawParams.replace(/^?/, '').split('&');
```

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
Thanks for this PR @Jefiozie - one small change I think.

Oh and you need to add a sentence in the body of the commit message:

> The commit message body does not meet the minimum length of 20 characters

File: packages/common/http/src/params.ts (hunk 1 above)
Comment (Jefiozie) [reply packages/common/http/src/params.ts:76]: You're totally right, thanks for the feedback will change it right away.

Comment (Jefiozie) [reply packages/common/http/src/params.ts:76]: Should the regex not be `/[\\^?]/`

Comment (petebacondarwin) [reply packages/common/http/src/params.ts]: Sorry it should be `/^\?/` (I forgot to escape the question mark.) This is using `^` here to mark the start of the string.
In `/[\\^?]/` this is saying match any character that is one of `\`, '^` or `?`.

**Comment by petebacondarwin** [comment]
I note that this is actually a duplicate of #30302. But since that PR has been awaiting updates from the author for some time, I am happy to close that and go with this one if we can get it in a mergeable state soonish.

File: packages/common/http/src/params.ts (hunk 1 above)
Comment (petebacondarwin) [reply packages/common/http/src/params.ts]: So just to be sure, add an extra test that checks that a question mark later in the param list is not replaced.

Comment (Jefiozie) [reply packages/common/http/src/params.ts]: Thank you for your explanation 🙏 pushing it while typing 😉:wq

**Comment by Jefiozie** [comment]
Sorry for that for some reason I didn't notice this. I'm happy to close this one in favor of the other one if that would be the right approach. I don't want to "grab" someone's PR.

**Comment by petebacondarwin** [comment]
Let's go with this PR. Can you add one more test that checks that we don't remove `?` from the rest of the param string? Thanks.

**Review by petebacondarwin — APPROVED** [review]
LGTM!  Thanks @Jefiozie

**Comment by Jefiozie** [comment]
Thank you for your help / guideance to merge this in.

**Review by alxhub — APPROVED** [review]

File: packages/common/http/test/params_spec.ts [hunk 2]
```diff
@@ -21,6 +21,18 @@ import {HttpParams} from '@angular/common/http/src/params';
...
       });
+
+      it('should ignore question mark in a url', () => {
+        const body = new HttpParams({fromString: '?a=b&c=d&c=e'});
+        expect(body.getAll('a')).toEqual(['b']);
+        expect(body.getAll('c')).toEqual(['d', 'e']);
+      });
+
+      it('should only remove question mark at the beginning of the params', () => {
+        const body = new HttpParams({fromString: '?a=b&c=d&?c=e'});
```
Comment (petebacondarwin) [inline packages/common/http/test/params_spec.ts:32]: The second expectation fails because there are now three separate parameter names: `a`, `c` and `?c`.

**Review by petebacondarwin — CHANGES_REQUESTED** [review]
Thanks for updating @Jefiozie.

There is a problem with the new test, and you will need to add a sentence to the body of the commit message to get it to pass the linting job on CI.

**Comment by Jefiozie** [comment]
Hi @petebacondarwin, thanks for you feedback. I've had some problems with running the test a couple of days. I've fixed the tests and added a body to the commit. Let me know if I need to change something still for this PR.

**Review by petebacondarwin — APPROVED** [review]
LGTM! Thanks @Jefiozie

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:357856886:BASE:357856891:1613532357385:7fe8f26c).

File: packages/common/http/src/params.ts [hunk 3]
```diff
@@ -73,7 +73,7 @@ export class HttpUrlEncodingCodec implements HttpParameterCodec {
 function paramParser(rawParams: string, codec: HttpParameterCodec): Map<string, string[]> {
   const map = new Map<string, string[]>();
   if (rawParams.length > 0) {
-    const params: string[] = rawParams.split('&');
+    const params: string[] = rawParams.replace(/^\?/, '').split('&');
```
Comment (AndrewKushnir) [inline packages/common/http/src/params.ts:76]: I'd propose adding a quick comment here explaining why we drop the leading `?` char:
```suggestion
    // The `window.location.search` can be used while creating an instance of the `HttpParams` class (e.g.
    // `new HttpParams({ fromString: window.location.search })`). The `window.location.search` may start with
    // the `?` char, so we strip it if it's present.
    const params: string[] = rawParams.replace(/^\?/, '').split('&');
```

**Review by AndrewKushnir — COMMENTED** [review]
@Jefiozie thanks for the PR. I've left a minor comment to add an inline doc to explain why we need to strip the `?` char.

Could you plz also update commit message (`git commit --amend`) to include `Fixes #28722` at the very end (so that the corresponding ticket is closed once this PR is merged)?

Thank you.

**Comment by AndrewKushnir** [comment]
@Jefiozie FYI internal tests in Google's codebase went well, so this PR should be ready to go.
Could you please have a look at https://github.com/angular/angular/pull/40610#discussion_r577301796 when you have a chance? Once it's addressed, we'll move this PR to the merge queue. Thank you.

**Comment by Jefiozie** [comment]
Hi @AndrewKushnir, thank you for the checks and feedback, I've made the changes you've requested. Hopefully, this can be merged in now.

**Comment by AndrewKushnir** [comment]
@Jefiozie great, thanks for the update.

I've set the `google3` status as passing and added the "merge" label. Once CI is done, this PR would be fully ready for merge.

Thanks for contributing to Angular!

**Pull request closed without merging**